### PR TITLE
Log collector broken pipe fix

### PR DESCRIPTION
**Pull request by kevinclark19a**
## Description

The agent service spawns a subprocess to build the log archive before sending the archive to wireserver itself.

If service shutdown occurs while that log collector subprocess is still executing, any attempts by that subprocess to write to stdout/stderr are met with a broken pipe exception (likely because the current code uses subprocess.Popen(..., stdout=subprocess.PIPE, stderr=subprocess.PIPE)).

This PR fixes this by instead sending the output to the agent's log file's descriptor (or, if some error occurs when attempting to get that descriptor, DEVNULL instead).

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: tests/ga/test_collect_logs.py [hunk 1]
```diff
@@ -120,6 +122,7 @@ def test_log_collection_is_invoked_with_resource_limits(self):
         self.assertIn("--property=CPUQuota=5%", args[0], "The log collector should have been invoked with a CPU limit")
         self.assertIn("--property=MemoryLimit=30M", args[0], "The log collector should have been invoked with a memory limit")
 
+    @attr('requires_sudo')
```
Comment (kevinclark19a) [inline tests/ga/test_collect_logs.py:125]: The logcollector handler thread now requires permissions to open the log file, and thus requires sudo.

File: azurelinuxagent/ga/collect_logs.py [hunk 2]
```diff
@@ -161,41 +161,53 @@ def _collect_logs():
...
-            msg = "Failed to collect logs. Elapsed time: {0} ms. Error: {1}".format(duration, exception_message)
-            # No need to log to the local log since we ran run_command with logging errors as enabled
-
-            return False
+            logfile = open(conf.get_agent_log_file(), "a+")
+        except Exception:
+            with subprocess_dev_null() as DEVNULL:
+                return exec_command(DEVNULL)
+        else:
+            return exec_command(logfile)
```
[the remark this answers is not in the document]
Comment (kevinclark19a) [reply azurelinuxagent/ga/collect_logs.py:208]: I had similar thoughts, but I think the `stdout`, `stderr` params for `subprocess.Popen` only accept integer file descriptors or file objects. I couldn't think of a way to hook up our entire `logger` module to this given those constraints, short of putting way too much work into it, like, e.g., basically recreating the file object class (or at least figuring out the methods that `Popen` uses on the file object internally and taking advantage of python's ducktyping to only implement those or something).
I didn't really want to go down that route, but if you have an idea for a reasonable implementation, I'd love to hear it.

As for the log file, currently all that's happening is we're dumping the `stdout` in directly, e.g.:

```
...
Running log collector mode normal
Log collection successfully completed. Archive can be found at /var/lib/waagent/logcollector/logs.zip and detailed log output can be found at /var/lib/waagent/logcollector/results.txt
...
```

File: azurelinuxagent/common/future.py [hunk 3]
```diff
@@ -34,6 +35,7 @@
     from builtins import int, range  # pylint: disable=unused-import,import-error
     from collections import OrderedDict  # pylint: disable=W0611
     from queue import Queue, Empty  # pylint: disable=W0611,import-error
+    import subprocess   # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (kevinclark19a) [reply azurelinuxagent/common/future.py:38]: The pylint error disable is actually for python2.7 pylint; it's used `if sys.version_info[0] == 3` (and is guarded as such above), but pylint isn't smart enough to figure that out for itself.

File: azurelinuxagent/ga/collect_logs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/collect_logs.py:208]: The output is decent enough for a good case (i.e. we only have stdout) but what if we have some error in the command and the stderr would just mess up the output of the log file. 

How about creating a temp file and capturing the output in that? Then at the end of it just read the output of the file and directly log it using the logger. 

Something like this - 
```python
with tempfile.TemporaryFile(dir=base_dir, mode="w+b") as output_file:
   exec_command(output_file)
```

File: tests/ga/test_collect_logs.py (hunk 1 above)
Comment (kevinclark19a) [reply tests/ga/test_collect_logs.py:125]: I personally didn't see this as a priority, but I do agree that a mock would be more ideal. If you think this is worth doing, I can add the change.

File: azurelinuxagent/ga/collect_logs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/collect_logs.py:208]: 1 more point to note, I see the `log_error=True` in the parameters for `run_command` - 

` shellutil.run_command(final_command, log_error=True, track_process=False,
                    stdout=output_file, stderr=output_file)`

You would also want to make sure we're not logging the error twice (once in run_command and the other time as part of your fix).

File: azurelinuxagent/common/future.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/future.py:38]: can you add a comment in the code explaining this? thanks

File: azurelinuxagent/common/future.py [hunk 4]
```diff
@@ -143,6 +145,21 @@ def is_file_not_found_error(exception):
...
+        # Suppress no-member errors on python2.7
+        yield subprocess.DEVNULL # pylint: disable=no-member
+    else:
+        try:
+            devnull = open(os.devnull, "a+")
+            yield devnull
+        except Exception:
+            yield None
+        finally:
+            devnull.close()
```
Comment (narrieta) [inline azurelinuxagent/common/future.py:161]: if open raises devnull will be None, so devnull.close() would raise

File: azurelinuxagent/common/future.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinclark19a) [reply azurelinuxagent/common/future.py:38]: will do-- I misread the comments above and thought they already mentioned that. Those comments are covering a slightly different scenario.

File: azurelinuxagent/ga/collect_logs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinclark19a) [reply azurelinuxagent/ga/collect_logs.py:208]: good catch, the double logging was in fact the case.

[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/collect_logs.py:208]: Synced offline about this, the logger stops working if the service is stopped. In this case logging directly to the logfile is the best way to capture all data. Resolving this comment

File: tests/ga/test_collect_logs.py [hunk 5]
```diff
@@ -77,6 +81,9 @@ def tearDown(self):
         if os.path.exists(self.archive_path):
             os.remove(self.archive_path)
         self.mock_archive_path.stop()
+        if os.path.exists(self.logger_path):
+            os.remove(self.logger_path)
+        self.mock_logger_path.stop()
```
Comment (larohra) [inline tests/ga/test_collect_logs.py:86]: You might need to stop the mock first and then delete file (can check it out)

**Review by dhivyaganesan — APPROVED** [review]

**Pull request merged**